64 bit build not possible with msys2 mingw64 and ASLR enabled

<!-- Please search existing issues for potential duplicates before filing yours:
https://github.com/godotengine/godot/issues?q=is%3Aissue
-->

**Godot version:**
<!-- Specify commit hash if using non-official build. -->
3.x 5e434841ec358ae3fefa2356b122c9d62df94179

**OS/device including version:**
Windows 10, msys2

**GCC:**
 gcc -v
Using built-in specs.
COLLECT_GCC=K:\dev\msys64\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=K:/dev/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-10.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts=yes --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --disable-plugin --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev9, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --with-boot-ldflags='-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high -Wl,--disable-dynamicbase -static-libstdc++ -static-libgcc' 'LDFLAGS_FOR_TARGET=-pipe -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high' --enable-linker-plugin-flags='LDFLAGS=-static-libstdc++\ -static-libgcc\ -pipe\ -Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high\ -Wl,--stack,12582912'
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (Rev9, Built by MSYS2 project)



**Issue description:**
It is not possible to build a 64-bit version with ASLR enabled by default.
see: https://www.msys2.org/news/

**errors:**
<pre>
modules/libmodules2.windows.opt.tools.64.a(gdscript.windows.opt.tools.64.o):gdscript.cpp:(.text$_ZN16GDScriptLanguage28debug_get_current_stack_infoEv[_ZN16GDScriptLanguage28debug_get_current_stack_infoEv]+0x20): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript.windows.opt.tools.64.o):gdscript.cpp:(.text$_ZN16GDScriptLanguage28debug_get_current_stack_infoEv[_ZN16GDScriptLanguage28debug_get_current_stack_infoEv]+0x30): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_editor.windows.opt.tools.64.o):gdscript_editor.cpp:(.text+0x4ae7): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_editor.windows.opt.tools.64.o):gdscript_editor.cpp:(.text+0x4af1): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_editor.windows.opt.tools.64.o):gdscript_editor.cpp:(.text+0x5b6a): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_editor.windows.opt.tools.64.o):gdscript_editor.cpp:(.text+0x5b74): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_function.windows.opt.tools.64.o):gdscript_function.cpp:(.text+0x1785): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_function.windows.opt.tools.64.o):gdscript_function.cpp:(.text+0x178f): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_function.windows.opt.tools.64.o):gdscript_function.cpp:(.text+0xea04): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_function.windows.opt.tools.64.o):gdscript_function.cpp:(.text+0xea0e): relocation truncated to fit: IMAGE_REL_AMD64_REL32 against undefined symbol `TLS init function for Thread::caller_id'
modules/libmodules2.windows.opt.tools.64.a(gdscript_function.windows.opt.tools.64.o):gdscript_function.cpp:(.text+0x2de18): additional relocation overflows omitted from the output
collect2.exe: error: ld returned 1 exit status\n"
=====
scons: [bin/godot.windows.opt.tools.64.exe] Error 1
scons: building terminated because of errors.
</pre>

**Steps to reproduce:**
mingw64 shell:
scons -j24 verbose=no use_mingw=yes bits=64 platform=windows target=release_debug debug_symbols=no module_bullet_enabled=no module_fbx_enabled=no module_gdnative_enabled=no tests=no

the same build params work fine with msvc without using msys2.

the only workaround I know is
LINKFLAGS=-Wl,--disable-dynamicbase,--default-image-base-low which unfortunately disables ASLR.

building only with LINKFLAGS=-Wl,--default-image-base-low builds too but the exe crashes at start with exception: 0xc0000005
error offset: 0x0000000000920000 

$ ./godot.windows.opt.tools.64.exe --verbose
WinTab context creation failed.
Godot Engine v3.3.rc.custom_build.5e434841e - https://godotengine.org
WinTab context creation failed.
Segmentation fault

PS
the prepackaged version from msys2 has ASLR disabled too according to procexp.

--- a/core/os/thread.cpp
+++ b/core/os/thread.cpp
@@ -47,7 +47,8 @@ uint64_t Thread::_thread_id_hash(const std::thread::id &p_t) {
 }
 
 Thread::ID Thread::main_thread_id = _thread_id_hash(std::this_thread::get_id());
-thread_local Thread::ID Thread::caller_id = 0;
+static thread_local Thread::ID caller_id = 0;
+static thread_local bool caller_id_cached = false;
 
 void Thread::_set_platform_funcs(
 		Error (*p_set_name_func)(const String &),
@@ -61,7 +62,9 @@ void Thread::_set_platform_funcs(
 }
 
 void Thread::callback(Thread *p_self, const Settings &p_settings, Callback p_callback, void *p_userdata) {
-	Thread::caller_id = _thread_id_hash(p_self->thread.get_id());
+	caller_id = _thread_id_hash(p_self->thread.get_id());
+	caller_id_cached = true;
+
 	if (set_priority_func) {
 		set_priority_func(p_settings.priority);
 	}
@@ -112,10 +115,6 @@ Error Thread::set_name(const String &p_name) {
 	return ERR_UNAVAILABLE;
 }
 
-Thread::Thread() {
-	caller_id = _thread_id_hash(std::this_thread::get_id());
-}
-
 Thread::~Thread() {
 	if (id != _thread_id_hash(std::thread::id())) {
 #ifdef DEBUG_ENABLED
@@ -125,4 +124,13 @@ Thread::~Thread() {
 	}
 }
 
+Thread::ID Thread::get_caller_id() {
+	if (likely(caller_id_cached)) {
+		return caller_id;
+	} else {
+		caller_id = _thread_id_hash(std::this_thread::get_id());
+		caller_id_cached = true;
+		return caller_id;
+	}
+}
 #endif

--- a/core/os/thread.h
+++ b/core/os/thread.h
@@ -66,7 +66,6 @@ private:
 	static uint64_t _thread_id_hash(const std::thread::id &p_t);
 
 	ID id = _thread_id_hash(std::thread::id());
-	static thread_local ID caller_id;
 	std::thread thread;
 
 	static void callback(Thread *p_self, const Settings &p_settings, Thread::Callback p_callback, void *p_userdata);
@@ -87,7 +86,7 @@ public:
 #if !defined(NO_THREADS)
 	_FORCE_INLINE_ ID get_id() const { return id; }
 	// get the ID of the caller thread
-	_FORCE_INLINE_ static ID get_caller_id() { return caller_id; }
+	static ID get_caller_id();
 	// get the ID of the main thread
 	_FORCE_INLINE_ static ID get_main_id() { return main_thread_id; }
 
@@ -98,7 +97,6 @@ public:
 	///< waits until thread is finished, and deallocates it.
 	void wait_to_finish();
 
-	Thread();
 	~Thread();
 #else
 	_FORCE_INLINE_ ID get_id() const { return 0; }

--- a/platform/windows/detect.py
+++ b/platform/windows/detect.py
@@ -397,6 +397,7 @@ def configure_mingw(env):
     ## Compile flags
 
     env.Append(CCFLAGS=["-mwindows"])
+    env.Append(LINKFLAGS=["-Wl,--nxcompat", "-Wl,--dynamicbase"])
     env.Append(CPPDEFINES=["WINDOWS_ENABLED", "OPENGL_ENABLED", "WASAPI_ENABLED", "WINMIDI_ENABLED"])
     env.Append(CPPDEFINES=[("WINVER", env["target_win_version"]), ("_WIN32_WINNT", env["target_win_version"])])
     env.Append(

--- a/platform/windows/godot_windows.cpp
+++ b/platform/windows/godot_windows.cpp
@@ -135,7 +135,7 @@ char *wc_to_utf8(const wchar_t *wc) {
 	return ubuf;
 }
 
-int widechar_main(int argc, wchar_t **argv) {
+__declspec(dllexport) int widechar_main(int argc, wchar_t **argv) {
 
 	OS_Windows os(NULL);
 
@@ -169,7 +169,7 @@ int widechar_main(int argc, wchar_t **argv) {
 	return os.get_exit_code();
 };
 
-int _main() {
+__declspec(dllexport) int _main() {
 	LPWSTR *wc_argv;
 	int argc;
 	int result;
@@ -187,7 +187,7 @@ int _main() {
 	return result;
 }
 
-int main(int _argc, char **_argv) {
+__declspec(dllexport) int main(int _argc, char **_argv) {
 	// _argc and _argv are ignored
 	// we are going to use the WideChar version of them instead
 

